### PR TITLE
Merge ssdk into main

### PR DESCRIPTION
**Pull request by JordonPhillips**
This takes the contents of the ssdk branch and updates it with main so that we can merge all the ssdk changes into main.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by adamthom-amzn — APPROVED** [review]
You gonna squash some/all of these?

**Comment by JordonPhillips** [comment]
Was not planning to squash, but I suppose we could. I'm 0 on it so vote here with reactions

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptDependency.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
     AWS_SDK_HASH_NODE("dependencies", "@aws-sdk/hash-node", "3.6.1", true),
 
     AWS_SDK_URL_PARSER("dependencies", "@aws-sdk/url-parser", "3.6.1", true),
+    AWS_SDK_URL_PARSER_NATIVE("dependencies", "@aws-sdk/url-parser-native", "3.6.1", true),
```
Comment (AllanZhengYP) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptDependency.java:47]: The package is committed mistakenly. The package has been removed in 3.13.0: https://github.com/aws/aws-sdk-js-v3/pull/2229

**Review by AllanZhengYP — COMMENTED** [review]
TypeScript part looks good to me, just have a view feedbacks/questions

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/CodegenVisitor.java [hunk 2]
```diff
@@ -175,11 +181,15 @@ void execute() {
...
         // files are switched between in package.json based on the targeted
         // environment.
-        RuntimeConfigGenerator configGenerator = new RuntimeConfigGenerator(
-                settings, model, symbolProvider, writers, integrations);
-        for (LanguageTarget target : LanguageTarget.values()) {
-            LOGGER.fine("Generating " + target + " runtime configuration");
-            configGenerator.generate(target);
+        if (settings.generateClient()) {
+            // For now these are only generated for clients.
+            // TODO: generate ssdk config
```
[the remark this answers is not in the document]
Comment (JordonPhillips) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/CodegenVisitor.java:186]: Something like that would be a great idea. I've got a PR that splits client/server up further where we could add something like that: https://github.com/awslabs/smithy-typescript/pull/314

File: smithy-typescript-ssdk-libs/server-common/src/httpbinding/mux.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    type: 'query_literal'
+    key: string
+    value?: string
+}
+
+export interface QuerySegment {
+    type: 'query'
+    key: string
+}
+export class UriSpec<S extends string, O extends String> {
```
[the remark this answers is not in the document]
Comment (JordonPhillips) [reply smithy-typescript-ssdk-libs/server-common/src/httpbinding/mux.ts]: 👍 I did a run of it over the code

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**